### PR TITLE
Various small fixes in crypto/{x509,bio}

### PR DESCRIPTION
**Pull request by DDvO**
* `X509_PUBKEY_set()`: Fix misleading error in case an arg is NULL
* `X509V3_EXT_CRL_add_nconf()`: Fix mem leak on error and simplify it
* `bio_lib.c`: Fix missing or misleading errors on NULL `BIO` argument

### CODE REVIEW

File: crypto/bio/bio_lib.c [hunk 1]
```diff
@@ -252,7 +252,11 @@ static int bio_read_intern(BIO *b, void *data, size_t dlen, size_t *readbytes)
 {
     int ret;
 
-    if ((b == NULL) || (b->method == NULL) || (b->method->bread == NULL)) {
+    if (b == NULL) {
+        ERR_raise(ERR_LIB_BIO, BIO_R_NULL_PARAMETER);
+        return -2;
```
Comment (levitte) [inline crypto/bio/bio_lib.c:257]: Please use `ERR_R_PASSED_NULL_PARAMETER`...  I really see no reason why we even have a separate BIO_R for exactly the same reason.
Also, please return -1 on this error, as -2 is "traditionally" used for unsupported methods.

**Review by levitte — CHANGES_REQUESTED** [review]

File: crypto/bio/bio_lib.c (hunk 1 above)
Comment (DDvO) [reply crypto/bio/bio_lib.c]: Good point.
When I did that change I was actually wondering why the BIO module (as well has some further ones)
have their own NULL pointer argument error (BTW, "argument" would be better than "parameter" here).
For consistency, I've replaced all `BIO_R_NULL_PARAMETER` by `ERR_R_PASSED_NULL_PARAMETER` (and purged the former).

Comment (DDvO) [reply crypto/bio/bio_lib.c]: I've also changed the return code in the above case.
Interesting to learn conventions about negative function result codes.
Are these documented/listed explicitly somewhere?

Comment (DDvO) [reply crypto/bio/bio_lib.c]: I've extended the list in #10725 accordingly: https://github.com/openssl/openssl/issues/10725#issuecomment-748820356

Comment (levitte) [reply crypto/bio/bio_lib.c]: Re documentation, the 1.1.1 `man BIO_read` says this in the RETURN VALUES section:

> All other functions return either the amount of data successfully read
> or written (if the return value is positive) or that no data was
> successfully read or written if the result is 0 or -1. If the return
> value is -2 then the operation is not implemented in the specific BIO
> type. ...

Comment (levitte) [reply crypto/bio/bio_lib.c]: > I've also changed the return code in the above case.

Why only this case?

Comment (DDvO) [reply crypto/bio/bio_lib.c]: Ah, I did not pay attention that the same issue was at other places in the same file.
I've improve them in the same way, and doing so also simplified various other return expressions in `bio_lib.c`.

File: crypto/x509/v3_conf.c [hunk 2]
```diff
@@ -365,17 +365,13 @@ int X509V3_EXT_CRL_add_nconf(CONF *conf, X509V3_CTX *ctx, const char *section,
...
+    int ret = X509V3_EXT_add_nconf_sk(conf, ctx, section, &exts);
 
-    if (req)
-        sk = &extlist;
-    i = X509V3_EXT_add_nconf_sk(conf, ctx, section, sk);
-    if (!i || !sk)
-        return i;
-    i = X509_REQ_add_extensions(req, extlist);
-    sk_X509_EXTENSION_pop_free(extlist, X509_EXTENSION_free);
-    return i;
```
Comment (t8m) [inline crypto/x509/v3_conf.c:378]: What is the supposed mem leak that is being fixed here? I do not see it. Also this breaks API as before the change X509V3_EXT_REQ_add_nconf was possible to be called with NULL req to just check the extensions without adding them to X509_REQ.

**Review by t8m — CHANGES_REQUESTED** [review]

File: include/openssl/bioerr.h [hunk 3]
```diff
@@ -106,7 +106,6 @@
 # define BIO_R_NO_HOSTNAME_OR_SERVICE_SPECIFIED           144
 # define BIO_R_NO_PORT_DEFINED                            113
 # define BIO_R_NO_SUCH_FILE                               128
-# define BIO_R_NULL_PARAMETER                             115
```
[the remark this answers is not in the document]
Comment (DDvO) [reply include/openssl/bioerr.h:109]: Well, so to be on the safe side (both technically and regarding a lengthy voting process)
I've re-added this `# define BIO_R_NULL_PARAMETER 115`, adding a comment that it is unused.
Such constants are hopefully not used externally, but who knows.

File: crypto/x509/v3_conf.c (hunk 2 above)
Comment (DDvO) [reply crypto/x509/v3_conf.c:378]: I had experienced a mem leak on failure of calling `X509V3_EXT_add_nconf_sk(conf, ctx, section, sk)`.
In this case the `sk` may be non-NULL and needs to be deallocated.

I was not aware of the 'just checking' use case; i've re-enabled it and added it to the documenting comment.
Moreover, I've added caveats to this and its closely related functions, such as:
`Note that on error new elements may remain added to req if req != NULL.`

**Comment by DDvO** [comment]
@t8m, I addressed your comments a week ago.
Can you please revisit this?

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged  - thanks @levitte and @t8m

**Pull request closed without merging**